custom_entity/Field/FieldWidget/DefaultEntityDurationWidget.php

```diff
@@ -0,0 +1,214 @@
+<?php
+
+namespace Drupal\custom_entity\Plugin\Field\FieldWidget;
+
+use Drupal\Core\Field\FieldItemListInterface;
+use Drupal\Core\Field\WidgetBase;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Ajax\AjaxResponse;
+use Drupal\Core\Ajax\HtmlCommand;
+
+/**
+ * Plugin implementation of the  'default_entity_duration_w' widget.
+ *
+ * @FieldWidget(
+ *   id = "default_entity_duration_w",
+ *   label = @Translation("Default Entity Duration - Widget"),
+ *   description = @Translation("Default Entity Duration - Widget"),
+ *   field_types = {
+ *     "default_entity_duration",
+ *   }
+ * )
+ */
+class DefaultEntityDurationWidget extends WidgetBase
+{
+  const INVALID_DURATION_ERROR_MESSAGE = 'End date must go after start date';
+
+  const INVALID_DATE_FORMAT_ERROR_MESSAGE = 'Incorrect date format';
+
+  /**
+   * {@inheritdoc}
+   */
+  public function formElement(
+    FieldItemListInterface $items,
+    $delta, array $element,
+    array &$form,
+    FormStateInterface $form_state
+  ) {
+    $type = $this->getSetting('date_format');
+
+    $element['start_date'] = [
+      '#type' => $type,
+      '#title' => $this->t('Start date'),
+      '#description' => $this->t('Select start date.'),
+      '#default_value' =>
+        $items[$delta]->start_date ?
+          date('Y-m-d', $items[$delta]->start_date) : '',
+      '#prefix' => '<span class="error-message"></span>',
+    ];
+
+    $element['end_date'] = [
+      '#type' => $type,
+      '#title' => t('End date'),
+      '#description' => t('Select end date'),
+      '#default_value' =>
+        $items[$delta]->end_date ?
+          date('Y-m-d', $items[$delta]->end_date) : '',
+      '#prefix' => '<span class="error-message"></span>',
+      '#element_validate' => [
+        [$this, 'validateDuration'],
+      ]
+    ];
+
+
+    if ($type != 'textfield')
+      return $element;
+
+    /** If type is textfield add ajax validation for elements */
+    $element['start_date']['#ajax'] = [
+      'callback' => [$this, 'validateAjax'],
+      'event' => 'change',
+    ];
+    array_unshift($element['start_date']['#element_validate'],
```

Comment:
Why you didn't add this validation above in the form element?

Reply:
Validation depends on the element type. In case it is a calendar we do not need the date format validation. The validation is necessary when dates are entered by keyboard.